packages/core/test/json.test.ts

```diff
@@ -0,0 +1,95 @@
+import { createMachine, assign } from '../src/index';
+import * as machineSchema from '../src/machine.schema.json';
+
+import * as Ajv from 'ajv';
+
+const ajv = new Ajv();
+const validate = ajv.compile(machineSchema);
+
+describe('json', () => {
+  it('should serialize the machine', () => {
+    const machine = createMachine<{ [key: string]: any }>({
+      initial: 'foo',
+      version: '1.0.0',
+      context: {
+        number: 0,
+        string: 'hello'
+      },
+      invoke: [{ id: 'invokeId', src: 'invokeSrc', autoForward: true }],
+      states: {
+        testActions: {
+          invoke: [{ id: 'invokeId', src: 'invokeSrc', autoForward: true }],
+          entry: [
+            'stringActionType',
+            {
+              type: 'objectActionType'
+            },
+            {
+              type: 'objectActionTypeWithExec',
+              exec: () => {
+                return true;
+              },
+              other: 'any'
+            },
+            function actionFunction() {
+              return true;
+            },
+            assign({
+              number: 10,
+              string: 'test',
+              evalNumber: () => 42
+            }),
+            assign(ctx => ({
+              ...ctx
+            }))
+          ],
+          on: {
+            TO_FOO: {
+              target: ['foo', 'bar'],
+              cond: ctx => !!ctx.string
+            }
+          }
+        },
+        foo: {},
+        bar: {},
+        testHistory: {
+          type: 'history',
+          history: 'deep'
+        },
+        testFinal: {
+          type: 'final',
+          data: {
+            something: 'else'
+          }
+        },
+        testParallel: {
+          type: 'parallel',
+          states: {
+            one: {
+              initial: 'inactive',
+              states: {
+                inactive: {}
+              }
+            },
+            two: {
+              initial: 'inactive',
+              states: {
+                inactive: {}
+              }
+            }
+          }
+        }
+      }
+    });
+
+    const json = JSON.parse(JSON.stringify(machine.definition));
+
+    try {
+      validate(json);
+    } catch (err) {
+      throw new Error(JSON.stringify(JSON.parse(err.message), null, 2));
+    }
+
+    expect(validate.errors).toBeNull();
```

Comment:
does this keep the latest errors on itself? 😱 what a... quirky API 🤣

Reply:
Yes, its pretty bad.